imbi/endpoints/project_sbom/models.py

```diff
@@ -0,0 +1,77 @@
+from __future__ import annotations
+
+import datetime
+import enum
+import typing
+
+import pydantic
+
+BOMRef = typing.Annotated[str, pydantic.StringConstraints(min_length=1)]
+PackageURL = typing.Annotated[str,
+                              pydantic.StringConstraints(pattern=r'^pkg:')]
+VersionMap = typing.MutableMapping[str, int]
+
+
+class Dependency(pydantic.BaseModel):
+    ref: BOMRef
+    depends_on: list[BOMRef] = pydantic.Field(default_factory=list,
+                                              alias='dependsOn')
+
+
+class ExternalReference(pydantic.BaseModel):
+    url: str
+    type: str
+
+
+class Component(pydantic.BaseModel):
+    name: str
+    bom_ref: typing.Union[BOMRef, None] = pydantic.Field(default=None,
+                                                         alias='bom-ref')
+    purl: typing.Union[PackageURL, None] = pydantic.Field(default=None,
+                                                          frozen=True)
+    version: typing.Union[str, None] = None
+    external_references: list[ExternalReference] = pydantic.Field(
+        default_factory=list, alias='externalReferences')
+
+    _package_purl: typing.Union[str, None] = None
+
+    @pydantic.field_validator('purl', mode='after')
+    @classmethod
+    def cache_package_purl(cls, value: str,
+                           info: pydantic.ValidationInfo) -> str:
+        _package_purl = value.partition('@')[0] if value else None
+        info.data['_package_purl'] = _package_purl
+        return value
+
+    @property
+    def package_purl(self) -> typing.Union[str, None]:
+        """Package URL without trailing info (eg, version, fragment, etc)"""
+        return self._package_purl
```

Comment:
I think using a [computed_field](https://docs.pydantic.dev/2.0/usage/computed_fields/) (potentially with a `cached_property`) makes more sense here

Reply:
hmm... I guess that I could do that since I froze the source data. Originally `package_purl` was a `cached_property` but there is no good way to clear the cache if you update the underlying field so I settled on the validation approach. It turns out that the validator doesn't work properly unless you run it on assignment ☹️ 

I can switch to a cached property again. I don't need it to be reserialized so I'm going to skip the `computed_field`. I don't serialize this model anywhere yet so it doesn't matter too much.